File: flutter/test/features/champion/data/repository/champion_repository_impl_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:lol_champion/core/error/app_exception.dart';
import 'package:lol_champion/features/champion/data/datasource/champion_remote_data_source.dart';
import 'package:lol_champion/features/champion/data/dto/champion_dto.dart';
import 'package:lol_champion/features/champion/data/repository/champion_repository_impl.dart';
import 'package:lol_champion/features/champion/domain/repository/patch_version_repository.dart';

void main() {
  group('ChampionRepositoryImpl', () {
    test('sorts the champion list by name', () async {
      final repository = ChampionRepositoryImpl(
        _FakeChampionRemoteDataSource(
          listResponse: ChampionResponseDto(
            version: 'test',
            data: {
              'Zed': _champion('Zed', '제드'),
              'Ahri': _champion('Ahri', '아리'),
            },
          ),
        ),
        _FakePatchVersionRepository(),
      );

      final champions = await repository.getChampionList();

      expect(champions.map((champion) => champion.name), ['아리', '제드']);
    });

    test('throws a domain error when a champion is missing', () async {
      final repository = ChampionRepositoryImpl(
        _FakeChampionRemoteDataSource(),
        _FakePatchVersionRepository(),
      );

      expect(
        () => repository.getChampion('Unknown'),
        throwsA(isA<ChampionNotFoundException>()),
      );
    });

    test('returns champion detail by id', () async {
      final repository = ChampionRepositoryImpl(
        _FakeChampionRemoteDataSource(
          detailResponse: ChampionResponseDto(
            version: 'test',
            data: {'Ahri': _champion('Ahri', '아리')},
          ),
        ),
        _FakePatchVersionRepository(),
      );

      final champion = await repository.getChampion('Ahri');

      expect(champion.id, 'Ahri');
      expect(champion.name, '아리');
    });
  });
}

ChampionDto _champion(String id, String name) => ChampionDto(
  id: id,
  name: name,
  title: '',
  lore: '',
  tags: const [],
  skins: const [],
  spells: const [],
);

class _FakeChampionRemoteDataSource implements ChampionRemoteDataSource {
  _FakeChampionRemoteDataSource({
    this.listResponse = const ChampionResponseDto(version: 'test', data: {}),
    this.detailResponse = const ChampionResponseDto(version: 'test', data: {}),
  });

  final ChampionResponseDto listResponse;
  final ChampionResponseDto detailResponse;

  @override
  Future<ChampionResponseDto> getChampionList(String patchVersion) async =>
      listResponse;

  @override
  Future<ChampionResponseDto> getChampion(
    String patchVersion,
    String championId,
  ) async => detailResponse;
}

class _FakePatchVersionRepository implements PatchVersionRepository {
  @override
  Future<String> getPatchVersion() async => 'test';

  @override
  Future<String> refreshLatestPatchVersion() async => 'test';
}

File: android/data/src/test/java/com/ex/lolcompose/data/repository/ChampionRepositoryImplTest.kt
package com.ex.lolcompose.data.repository

import com.ex.lolcompose.data.datasource.ChampionRemoteDataSource
import com.ex.lolcompose.domain.common.DataResult
import com.ex.lolcompose.domain.model.Champion
import com.ex.lolcompose.domain.model.ChampionInfo
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.IOException

class ChampionRepositoryImplTest {

    @Test
    fun `champion list is sorted by name`() = runBlocking {
        val dataSource = FakeChampionRemoteDataSource(
            listResponse = championInfo(
                Champion(id = "Zed", name = "제드"),
                Champion(id = "Ahri", name = "아리")
            )
        )

        val result = ChampionRepositoryImpl(dataSource).getChampionList().first()

        assertTrue(result is DataResult.Success)
        val champions = (result as DataResult.Success).data
        assertEquals(listOf("아리", "제드"), champions.map { it.name })
    }

    @Test
    fun `network exception is returned as an error`() = runBlocking {
        val exception = IOException("network unavailable")
        val dataSource = FakeChampionRemoteDataSource(listException = exception)

        val result = ChampionRepositoryImpl(dataSource).getChampionList().first()

        assertTrue(result is DataResult.Error)
        assertEquals(exception, (result as DataResult.Error).exception)
    }

    @Test
    fun `missing champion is returned as an error`() = runBlocking {
        val dataSource = FakeChampionRemoteDataSource(detailResponse = championInfo())

        val result = ChampionRepositoryImpl(dataSource).getChampion("Unknown").first()

        assertTrue(result is DataResult.Error)
        assertTrue((result as DataResult.Error).exception is NoSuchElementException)
    }

    private fun championInfo(vararg champions: Champion): ChampionInfo<Champion> = ChampionInfo(
        type = "champion",
        version = "test",
        data = champions.associateByTo(mutableMapOf()) { it.id }
    )

    private class FakeChampionRemoteDataSource(
        private val listResponse: ChampionInfo<Champion> = ChampionInfo(
            type = "champion",
            version = "test",
            data = mutableMapOf()
        ),
        private val detailResponse: ChampionInfo<Champion> = listResponse,
        private val listException: Throwable? = null
    ) : ChampionRemoteDataSource {
        override suspend fun getChampionList(): ChampionInfo<Champion> {
            listException?.let { throw it }
            return listResponse
        }

        override suspend fun getChampion(id: String): ChampionInfo<Champion> = detailResponse
    }
}
